src/src/WebServer/JSON.cpp

```diff
@@ -448,6 +448,12 @@ void handle_json()
         stream_next_json_object_value(F("Type"),             getPluginNameFromDeviceIndex(DeviceIndex));
         stream_next_json_object_value(F("TaskName"),         getTaskDeviceName(TaskIndex));
         stream_next_json_object_value(F("TaskDeviceNumber"), Settings.TaskDeviceNumber[TaskIndex]);
+        for(int i = 0; i < 3; i++) {
+          if (Settings.TaskDevicePin[i][TaskIndex] != -1) {
```

Comment:
Maybe best to use the `Device[DeviceIndex].usesTaskDevicePin(i)` function here. (next to checking for the pin >= 0)
This also requires you to fetch the `DeviceIndex` (outside the loop please ;) ).
This way you don't output a value from a previous setting or some perhaps uninitialized value.

Reply:
> Maybe best to use the Device[DeviceIndex].usesTaskDevicePin(i) function here. (next to checking for the pin >= 0)

I don´t really get why i would check  `pin >= 0` when `usesTaskDevicePin(i)` is bool...

Is this what you had in mind?

```cpp
        int devIndex = DeviceIndex;
        for(int i = 1; i < 4; i++) {
          if (Device[devIndex].usesTaskDevicePin(i)) {
            stream_next_json_object_value(concat(F("TaskDeviceGPIO"), i) , String(Settings.TaskDevicePin[i-1][TaskIndex]));
          }
        }
```

Reply:
Nope :)

> I don´t really get why i would check `pin >= 0` when `usesTaskDevicePin(i)` is bool...

`usesTaskDevicePin` does only check the type (e.g. `DEVICE_TYPE_SINGLE`)
Now imagine the use case where the storage used for a task was initialized to `0` but not set the `TaskDevicePin` to `-` when not used. (which strictly speaking might be considered a bug if it occurs)
Then `Settings.TaskDevicePin[i][TaskIndex]` might be `!= -1` for `i > 0` and `type == DEVICE_TYPE_SINGLE`.

This extra check makes sure to not even allow those "TaskDeviceGPIO" values to be output in the JSON when they should not be used.

```c++
        for(int i = 0; i < 3; i++) {
          if (Settings.TaskDevicePin[i][TaskIndex] != -1 && Device[DeviceIndex].usesTaskDevicePin(i + 1)) {
            stream_next_json_object_value(concat(F("TaskDeviceGPIO"), i + 1) , String(Settings.TaskDevicePin[i][TaskIndex]));
          }
        }
```

One tip:
Better have the loop start at 0 and add one, compared to starting at 1 and subtract one.
Subtraction from an int type always makes me nervous, expecially when used as an index in some array.
Better correct this `i` value where it isn't used such that it doesn't start at 0. Thus the exception needs correcting, where "normal use" (thus starting at 0) does not need correcting.

Reply:
The issue with this code, that for an i2c device like an oled where you can assign a third pin as a button it seems that `usesTaskDevicePin(3)` is also false.

<img width="461" alt="Bildschirmfoto 2022-12-07 um 10 49 54" src="https://user-images.githubusercontent.com/33860956/206146001-c1ad102f-4b93-4bf2-9846-7e391f648b0c.png">


| Output with the code of this PR | Output with the code above |
| ------------- | ------------- |
|  <img width="237" alt="Bildschirmfoto 2022-12-07 um 10 50 34" src="https://user-images.githubusercontent.com/33860956/206146106-eb41fa88-2926-4538-a2a2-1ea41a6683ed.png"> | <img width="270" alt="Bildschirmfoto 2022-12-07 um 10 30 05" src="https://user-images.githubusercontent.com/33860956/206145531-c049aa6d-30f1-4a14-bb00-750eefec71f2.png">  |

Reply:
There's this [FIXME](https://github.com/letscontrolit/ESPEasy/blob/mega/src/_P036_FrameOLED.ino#L301) in the source, but also this [SENSOR_TYPE_NONE](https://github.com/letscontrolit/ESPEasy/blob/mega/src/_P036_FrameOLED.ino#L206) setting for the device... So the check is working as intended, but it's not what you expected or desired 😅

Reply:
> The issue with this code, that for an i2c device like an oled where you can assign a third pin as a button it seems that usesTaskDevicePin(3) is also false.

Ohh, i answered to the first post of you und didn’t realize, that you already had a new suggestion..  I need to check this

Reply:
And also, many plugins (also) use regular `PCONFIG(n)` settings to configure GPIO pins. And some use `PLUGIN_WEBFORM_SHOW_GPIO_DESCR` function to provide that info to the Devices overview, but I'm not sure if that should be included in the json output though, as it can contain formatting, and not all plugins use that function.

Reply:
Ok, so what does it mean for this PR. Leave it as it is right now, since it displays the GPIOs correctly in JSON, or go with the latest suggestion of @TD-er but having a somewhat inconsistent JSON output?
Or (my least favorite) waiting with this PR until everything meets a certain standard...

Reply:
Let's "assume" there are no other bugs in clearing/initializing the `TaskDevicePin` values and then we can keep what you already had.
Let's not wait for other unknown (and undefined) "whatifs" and merge it...
@tonhuisman you agree?

Reply:
I'd leave it for now to only check for the pins to be >= 0. Once the other issues are addressed, the JSON output should maybe be reconsidered/updated.